examples/fedscr/fedscr_server.py

```diff
@@ -0,0 +1,201 @@
+            for name, weights in deltas_received[0].items()
+        }
+
+        for i, update in enumerate(deltas_received):
+            __, report, __, __ = updates[i]
+            num_samples = report.num_samples
+
+            for name, delta in update.items():
+                # Use weighted average by the number of samples
+                avg_update[name] += delta * (num_samples / self.total_samples)
+
+            # Yield to other tasks in the server
+            await asyncio.sleep(0)
+
+        return avg_update
+
+    async def process_reports(self):
+        """Process the client reports by aggregating their weights."""
+        await self.aggregate_weights(self.updates)
+        config = Config().trainer._asdict()
+
+        # Testing the global model accuracy
+        if hasattr(Config().server, "do_test") and not Config().server.do_test:
+            # Compute the average accuracy from client reports
+            self.accuracy = self.accuracy_averaging(self.updates)
+            logging.info(
+                "[%s] Average client accuracy: %.2f%%.", self, 100 * self.accuracy
+            )
+        else:
+            # Testing the updated model directly at the server
+            self.accuracy = self.trainer.test_model(
+                config, self.testset, self.testset_sampler
+            )
+
+        if hasattr(Config().trainer, "target_perplexity"):
+            logging.info("[%s] Global model perplexity: %.2f\n", self, self.accuracy)
+        else:
+            logging.info(
+                "[%s] Global model accuracy: %.2f%%\n", self, 100 * self.accuracy
+            )
+
+        if self.trainer.use_adaptive:
+            self.mean_variance = self.calc_loss_var()
+
+    def calc_loss_var(self):
+        """Calculate the loss variance using mean squared error."""
+        global_loss = [sum(self.local_loss) / len(self.local_loss)]
+        loss = torch.nn.MSELoss()
+        variance = loss(
+            torch.FloatTensor(self.local_loss), torch.FloatTensor(global_loss)
+        )
+        self.variances.append(variance.data.item())
+
+        mew = sum(variance for variance in self.variances)
+        if self.current_round > 3:
+            mew = mew * (1 / (self.current_round - 2))
+        else:
+            mew = 0
+        return mew
+
+    async def select_clients(self, for_next_batch=False):
+        """Send the update_thresholds to the clients."""
+        await super().select_clients(for_next_batch)
+
+        if self.trainer.use_adaptive:
+            self.calc_threshold()
+            checkpoint_path = Config().params["checkpoint_path"]
+            model_name = Config().trainer.model_name
+
+            if not os.path.exists(checkpoint_path):
+                os.makedirs(checkpoint_path)
+
+            loss_path = f"{checkpoint_path}/{model_name}_thresholds.pkl"
+
+            with open(loss_path, "wb") as file:
+                pickle.dump(self.update_threshold, file)
+
```

Comment:
The way that FedSCR implements passing information (thresholds in this case) from server to the clients is completely wrong. It used files, which must be changed to using the communication channel.

Reply:
Changed in commit 63b2bb9ed59f74680d1d7028943363636f48c0d0.